src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonDerivedType.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    /// <summary>
+    /// Represents a supported derived type defined in the metadata of a polymorphic type.
+    /// </summary>
+    public readonly struct JsonDerivedType
+    {
+        /// <summary>
+        /// Specifies a supported derived type without a type discriminator.
+        /// </summary>
+        /// <param name="derivedType">The derived type to be supported by the polymorphic type metadata.</param>
+        public JsonDerivedType(Type derivedType)
+        {
+            DerivedType = derivedType;
+            TypeDiscriminator = null;
+        }
+
+        /// <summary>
+        /// Specifies a supported derived type with an integer type discriminator.
+        /// </summary>
+        /// <param name="derivedType">The derived type to be supported by the polymorphic type metadata.</param>
+        /// <param name="typeDiscriminator">The type discriminator to be associated with the derived type.</param>
+        public JsonDerivedType(Type derivedType, int typeDiscriminator)
+        {
+            DerivedType = derivedType;
+            TypeDiscriminator = typeDiscriminator;
+        }
+
+        /// <summary>
+        /// Specifies a supported derived type with a string type discriminator.
+        /// </summary>
+        /// <param name="derivedType">The derived type to be supported by the polymorphic type metadata.</param>
+        /// <param name="typeDiscriminator">The type discriminator to be associated with the derived type.</param>
+        public JsonDerivedType(Type derivedType, string typeDiscriminator)
+        {
+            DerivedType = derivedType;
+            TypeDiscriminator = typeDiscriminator;
```

Comment:
(although IMO it should just be nullable and equivalent to the other ctor without it)

Reply:
Intentionally left out null validation here since it's a struct and can be set to null via the default constructor. Since it's already being validated at Configured figured this should be a simple value class without validation logic but the annotation should still guide the user to not pass null types.